app/code/Magento/ImportExport/Model/Import/Source/FileParser/CsvParser.php

```diff
@@ -0,0 +1,113 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\ImportExport\Model\Import\Source\FileParser;
+
+use Magento\Framework\App\Filesystem\DirectoryList;
+use Magento\Framework\Filesystem;
+
+class CsvParser implements ParserInterface
+{
+    /**
+     * List of CSV parsing options
+     *
+     * @var array
+     */
+    private $options = [
+        'delimiter' => ',',
+        'enclosure' => '"',
+        'escape' => '\\',
+        'null' => null
+    ];
+
+    /**
+     * File reader
+     *
+     * @var Filesystem\File\ReadInterface
+     */
+    private $file;
+
+    /**
+     * Columns of CSV file
+     *
+     * @var string[]
+     */
+    private $columns;
+
+    public function __construct(
+        Filesystem\File\ReadInterface $file,
+        $options = []
+    ) {
+        $this->file = $file;
+        $this->options = $options + $this->options;
+        $this->columns = $this->fetchCsvLine();
+
+        if ($this->columns === false) {
+            throw new \InvalidArgumentException('CSV file should contain at least 1 row');
+        }
+    }
+
+    public function getColumnNames()
+    {
+        return $this->columns;
```

Comment:
It has a client `FileSource` that requires column names for proper Source implementation.